lib/src/lifecycle_module.dart

```diff
@@ -207,45 +267,80 @@ abstract class LifecycleModule extends Object with Disposable {
   /// Suspend indicates to the module that it should go into a low-activity
   /// state. For example, by disconnecting from backend services and unloading
   /// heavy data structures.
-  Future suspend() {
-    Completer completer = new Completer();
-    if (!_isSuspended && _isLoaded) {
-      _willSuspendController.add(this);
-      Future.wait(_childModules.map((c) => c.suspend())).then((_) {
-        onSuspend().then((_) {
-          _didSuspendController.add(this);
-          _isSuspended = true;
-          completer.complete();
-        });
-      });
-    } else {
-      _logger.warning(
-          'Module "$name" is ${_isLoaded ? 'already suspended' : 'not loaded'}, cannot suspend.');
-      completer.complete();
+  ///
+  /// Initiates the suspend process when the module is in the loaded state. If
+  /// the module is in the suspended or suspending state a warning is logged and
+  /// the method is a noop. If the module is in any other state, a StateError is
+  /// thrown.
+  Future<Null> suspend() {
+    if (isSuspended || isSuspending) {
+      return _buildNoopResponse(
+          isTransitioning: isSuspending,
+          methodName: 'suspend',
+          currentState: LifecycleState.suspended);
     }
+
+    if (!isLoaded) {
+      return _buildIllegalTransitionResponse(
+          targetState: LifecycleState.suspended,
+          allowedStates: [LifecycleState.loaded]);
+    }
+
+    _state = LifecycleState.suspending;
+    final completer = new Completer<Null>();
+    _transitionFuture = completer.future;
+    _willSuspendController.add(this);
+
+    Future.wait(_childModules.map((c) => c.suspend())).then((_) {
+      onSuspend().then((_) {
+        _didSuspendController.add(this);
+        _state = LifecycleState.suspended;
```

Comment:
#nit Can we set the state first? Ultimately it won't matter but just to keep things consistent and clearly express what we're trying to do it makes sense in my mind.